File: android/app/src/main/java/com/izzatismail/midtrim/domain/usecase/ReorderVideosUseCase.kt
package com.izzatismail.midtrim.domain.usecase

import com.izzatismail.midtrim.domain.error.ReorderVideosError

class ReorderVideosUseCase {
    fun <T> execute(items: List<T>, newOrder: List<Int>): List<T> {
        if (items.isEmpty()) throw ReorderVideosError.EmptyList
        if (newOrder.size != items.size || newOrder.toSet() != (0 until items.size).toSet()) {
            throw ReorderVideosError.InvalidIndices
        }
        return newOrder.map { items[it] }
    }
}

File: ios/MidTrim/MidTrim/Domain/UseCases/ReorderVideosUseCase.swift
import Foundation

struct ReorderVideosUseCase {
    func execute<T>(_ items: [T], to newOrder: [Int]) throws -> [T] {
        guard !items.isEmpty else { throw ReorderVideosError.emptyList }
        guard newOrder.count == items.count,
              Set(newOrder) == Set(0..<items.count) else {
            throw ReorderVideosError.invalidIndices
        }
        return newOrder.map { items[$0] }
    }
}
